File: python/Math&Geometry/detect_squares.py
from typing import List
from collections import defaultdict

class CountSquares:
    """
    Data structure to add points on a 2D plane and count the number of axis-aligned squares
    that can be formed with a given query point as one of the corners.
    
    Approach:
    - Use a dictionary (ptsCount) to count occurrences of each point (for duplicate points)
    - Store all added points in a list (pts) for iteration
    - For a query point (px, py), iterate over all previously added points (x, y)
      and check if they can form a diagonal of a square with (px, py)
    - For each such diagonal, check if the other two corners exist using ptsCount
    - Multiply the counts of the other two corners to get the number of squares for this diagonal
    
    Time Complexity:
    - add: O(1) per operation
    - count: O(N), where N is the number of points added so far
      (since we iterate over all points for each count query)
    
    Space Complexity: O(N), where N is the number of points added (for pts and ptsCount)
    """
    def __init__(self):
        self.ptsCount = defaultdict(int)  # Maps (x, y) -> count
        self.pts = []  # List of all points added

    def add(self, point: List[int]) -> None:
        """
        Add a point to the data structure.
        Args:
            point: List[int] - The [x, y] coordinates of the point
        """
        self.ptsCount[tuple(point)] += 1
        self.pts.append(point)

    def count(self, point: List[int]) -> int:
        """
        Count the number of axis-aligned squares with the given point as one of the corners.
        Args:
            point: List[int] - The [x, y] coordinates of the query point
        Returns:
            int - The number of squares
        """
        res = 0
        px, py = point

        for x, y in self.pts:
            # Check if (x, y) can be a diagonal with (px, py) (must form a square)
            if (abs(py - y) != abs(px - x)) or x == px or y == py:
                continue
            # The other two corners are (x, py) and (px, y)
            res += self.ptsCount[(x, py)] * self.ptsCount[(px, y)]
        return res

# -------------------
# Test Cases
# -------------------
def test_detect_squares():
    cs = CountSquares()
    # Add points
    cs.add([3, 10])
    cs.add([11, 2])
    cs.add([3, 2])
    # Test case 1: No square yet
    assert cs.count([11, 10]) == 1, "Failed on count([11, 10])"
    # Add more points
    cs.add([11, 2])
    # Test case 2: Duplicate point increases square count
    assert cs.count([11, 10]) == 2, "Failed on count([11, 10]) after duplicate add"
    # Test case 3: No square possible
    assert cs.count([14, 8]) == 0, "Failed on count([14, 8])"
    # Add more points to form a square at a different location
    cs2 = CountSquares()
    cs2.add([1, 1])
    cs2.add([1, 2])
    cs2.add([2, 1])
    cs2.add([2, 2])
    assert cs2.count([1, 1]) == 1, "Failed on count([1, 1]) for perfect square"
    assert cs2.count([2, 2]) == 1, "Failed on count([2, 2]) for perfect square"
    # Test case 4: Multiple squares with shared points
    cs3 = CountSquares()
    cs3.add([0, 0])
    cs3.add([0, 1])
    cs3.add([1, 0])
    cs3.add([1, 1])
    cs3.add([2, 0])
    cs3.add([2, 1])
    assert cs3.count([1, 0]) == 2, "Failed on count([1, 0]) for two possible squares"
    print("All test cases passed!")

if __name__ == "__main__":
    test_detect_squares()

File: C++/Math&Geometry/detect_squares.cpp
#include <bits/stdc++.h>
using namespace std;

class CountSquares {
private:
    unordered_map<string, int> ptsCount; // maps "x,y" -> count
    vector<pair<int, int>> pts;          // stores all points

    // Helper to generate a string key from x and y
    string key(int x, int y) {
        return to_string(x) + "," + to_string(y);
    }

public:
    /*
     * Constructor: Initializes the data structure.
     */
    CountSquares() {
        // No additional initialization required
    }

    /*
     * Add a point to the data structure.
     * Time Complexity: O(1)
     */
    void add(vector<int> point) {
        int x = point[0], y = point[1];
        ptsCount[key(x, y)]++;       // increment count of (x, y)
        pts.emplace_back(x, y);     // store point
    }

    /*
     * Count the number of axis-aligned squares with given point as one corner.
     * Time Complexity: O(N), where N = number of points added.
     */
    int count(vector<int> point) {
        int res = 0;
        int px = point[0], py = point[1];

        for (auto& p : pts) {
            int x = p.first, y = p.second;

            // Skip if not forming a square diagonal:
            if (abs(py - y) != abs(px - x) || x == px || y == py) continue;

            // Other two corners: (x, py) and (px, y)
            res += ptsCount[key(x, py)] * ptsCount[key(px, y)];
        }
        return res;
    }
};

/*
 * 📖 Line-by-line explanation:
 *
 * #include <bits/stdc++.h>           → Includes all standard C++ libraries.
 * using namespace std;              → Allows standard names without std:: prefix.
 *
 * class CountSquares { … };        → Defines a class named CountSquares.
 *
 * private:                          → Members declared private to hide implementation.
 *
 * unordered_map<string, int> ptsCount; → Stores counts of each point, key as "x,y".
 * vector<pair<int, int>> pts;           → Stores list of all points added so far.
 *
 * string key(int x, int y) { … }   → Helper to create a unique string for a point.
 *
 * public:                          → Public interface of the class.
 *
 * CountSquares() { … }            → Constructor — initializes an empty object.
 *
 * void add(vector<int> point) { … } → Adds a point to the data structure.
 * int x = point[0], y = point[1];  → Extract x and y from the vector.
 * ptsCount[key(x,y)]++;           → Increment count of that point.
 * pts.emplace_back(x,y);          → Add to the list of points.
 *
 * int count(vector<int> point) { … } → Counts the number of squares.
 * int res = 0;                      → Initialize result counter.
 * for (auto& p : pts)              → Iterate over all previously added points.
 * int x = p.first, y = p.second;  → Get x and y from pair.
 *
 * if (abs(py-y) != abs(px-x) || x==px || y==py) continue; → Skip invalid diagonals.
 *
 * res += ptsCount[key(x,py)] * ptsCount[key(px,y)]; → Multiply counts of other corners.
 *
 * return res;                       → Return total number of squares.
 *
 */

